### ISSUE TITLE
hope generate more method with 'MyBatis3DynamicSql' model

### ISSUE THREAD

**Issue by pky920216344** [mybatis/generator#472]
when i use method  like  xxxxByExample in xxxMapper, must  coding 'build().execute()' in the end every time. so i think why not like Mybatis-Spring, Mybatis-Spring will auto close SqlSession.
here are some of my thoughts

1. In mybatis-dynamic-sql project, where DSL, Update DSL, Update DSL. Update Where Builder, WhereDSL and so on can implement the interface of org.mybatis.dynamic.sql.util.Buildable.
2. generate more method like this:

```java
    default List<Model> selectWithFunction(
Function<QueryExpressionDSL<MyBatis3SelectModelAdapter<List<Model>>>,
            Buildable<MyBatis3SelectModelAdapter<List<Model>>>> function) {
        return function.apply(this.selectByExample()).build().execute();
    }
```

**Comment by jeffgbutler** [comment]
Another interesting idea from @pky920216344!

I experimented with this a bit. I wrote a mapper method like this:

```java
default List<SimpleTableRecord> selectByExample(
        Function<QueryExpressionDSL<MyBatis3SelectModelAdapter<List<SimpleTableRecord>>>,
        Buildable<MyBatis3SelectModelAdapter<List<SimpleTableRecord>>>> function) {
    return function.apply(selectByExample()).build().execute();
}
```

Note that I called it selectByExample again. Then it can be used like this:

```java
List<SimpleTableRecord> rows = mapper.selectByExample(q -> q
        .where(id, isEqualTo(1))
        .or(occupation, isNull())
        .limit(3));
```

The mapper method has an ugly method signature for sure, but it enables better looking client code.

As you say, we could build methods like this for selectByExample, deleteByExample, updateByExample, etc.

I'm not sure if we should build the methods as a normal part of code generation, or build them with an optional plugin. What do you think?

**Comment by jeffgbutler** [comment]
I think there are three options:

1. Generate these new methods and the old methods as a part of the normal code generation process
1. Generate the old methods by default and generate the new methods in a plugin - this is least disruptive
1. Generate the new methods by default and move the old methods to a "legacy methods" plugin - this is disruptive, but probably a better long term solution

I'm leaning towards the last option. Doing that will break some existing generated code if anyone runs the generator again, but we could probably make that clear in release notes.

**Comment by pky920216344** [comment]
as you say, this may just be for my project.
perhaps the old methods and new methods can be retained at the same time, and there may be better solutions in the future.

**Comment by jeffgbutler** [comment]
@pky920216344 Have a look at this commit...https://github.com/jeffgbutler/mybatis-dynamic-sql/commit/5b3be4b1c24942adf50f796389e1dd454b5ef5a9

I added some utility interfaces to make the mapper method signatures less ugly, and then wrote an example mapper and a bunch of tests. I'm still thinking about the best way to introduce this change, but I think I like it. I'm interested in any feedback.

**Comment by pky920216344** [comment]
@jeffgbutler  look so good 😀

**Comment by jeffgbutler** [comment]
Resolved via #487
